### PR TITLE
week-5

### PR DESCRIPTION
**Pull request by oleg3555**
Now in app you can :

1. check tasks of selected user
2. see user info in members table
3. get data from firestore
4. check progress page
5. check tasks page
6. delete users,tasks,tracks
7. add tasks, tracks
8. edit tasks and tracks
9. read tasks and tracks in modals
10. look subtasks of every user
11. login with email and password
12. send email with link so set password( like custom sign up)
13. add delete and edit user in modal
14. some new features

[Link to deploy](https://happy-benz-165b32.netlify.app/users)

### CODE REVIEW

File: src/pages/modals/userModal/helperTests/getReadUserModalState.test.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      surname: 'Reduxov',
+    };
+
+    const actual = getEditUserModalState(user);
+    expect(actual).toEqual(result);
+  });
+
+  it('Should return wrong data to read in modal', () => {
+    const user = {
+      address: 'Руссиянова 4',
```
Comment (maximsan) [inline src/pages/modals/userModal/helperTests/getReadUserModalState.test.js:71]: Don't use Russian with English. Use only English !

File: src/pages/modals/userModal/userModalHelpers.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  address: '',
+  phone: '',
+  skype: '',
+  startDate: '',
+  education: '',
+  averageScore: '',
+  mathScore: '',
+};
+
+export const initStartUserModalState = {
```
Comment (maximsan) [inline src/pages/modals/userModal/userModalHelpers.js:24]: name it `userState` or `userModalState`

File: src/pages/modals/userModal/userModalHelpers.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  return { ...initStartUserModalState, ...user, modalTitle: USER_MODAL_TITLES.read, readOnly: true };
+};
+
+export const getEditUserModalState = (user) => {
+  return { ...initStartUserModalState, ...user, modalTitle: USER_MODAL_TITLES.edit, readOnly: false };
+};
+
+export const getCreateUserModalState = () => {
+  return { ...initStartUserModalState, modalTitle: USER_MODAL_TITLES.create, readOnly: false };
+};
```
Comment (maximsan) [inline src/pages/modals/userModal/userModalHelpers.js:78]: export const getUserModalState = (title, readOnly = false) => {
  return { ...initStartUserModalState, title, readOnly };
};
`getUserModalState(user, USER_MODAL_TITLES.edit)`

File: src/pages/setPassword/SetPassword.jsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  changePasswordValue = (event) => {
+    this.setState({ password: event.currentTarget.value, error: false });
+  };
+
+  setPassword = async () => {
+    const { password, user } = this.state;
+    const { history } = this.props;
+    if (password.length < 8 || password.length > 24) {
```
Comment (maximsan) [inline src/pages/setPassword/SetPassword.jsx:43]: reuse this code

File: src/pages/setPassword/SetPassword.jsx [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      } catch (error) {
+        console.log(error);
+      }
+    }
+  };
+
+  render() {
+    const { user, password, error } = this.state;
+    if (!user) {
+      return <div className={styles.wrapper}>Loading...</div>;
```
Comment (maximsan) [inline src/pages/setPassword/SetPassword.jsx:59]: create <Loader /> component

File: src/pages/setPassword/SetPassword.jsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+          <div className={styles.title}>Set your own password</div>
+          <div className={styles.inputs}>
+            <input type={INPUT_TYPES.text} value={user.email} disabled />
+            <input
+              type={INPUT_TYPES.password}
+              value={password}
+              placeholder='Password'
+              onChange={this.changePasswordValue}
+            />
+            {error && <div className={styles.error}>Password should contains 8-24 symbols</div>}
```
Comment (maximsan) [inline src/pages/setPassword/SetPassword.jsx:74]: create `<Error />` component

File: src/pages/tasks/Tasks.jsx [hunk 7]
```diff
@@ -49,10 +49,15 @@ export class Tasks extends PureComponent {
...
-      const { users } = updatedTask;
-      const prevUsers = prevTask.users;
-      await updateTask(actionTaskId, updatedTask);
+    const { users } = updatedTask;
+    const prevUsers = prevTask.users;
+    const updatedUsers = users.map((user) => {
+      const item = prevUsers.find((el) => el.userId === user.userId);
+
+      return item ? { ...user, status: item.status } : user;
+    });
```
Comment (maximsan) [inline src/pages/tasks/Tasks.jsx:58]: const newTask = { ...updatedTask, users: updatedUsers }

File: src/tests/getAge.test.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import { getAge } from '../scripts/helpers';
+
+describe('Helper functions tests', () => {
```
Comment (maximsan) [inline src/tests/getAge.test.js:3]: `getAge`

File: src/tests/getUserName.test.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import { getUserName } from '../scripts/helpers';
+
+describe('Helper functions tests', () => {
```
Comment (maximsan) [inline src/tests/getUserName.test.js:3]: `getUserName`

File: src/pages/members/Members.jsx [hunk 10]
```diff
@@ -43,34 +47,74 @@ export class Members extends PureComponent {
...
 
   render() {
-    const { users, deleteMode } = this.state;
+    const { users, modalMode, actionUserId } = this.state;
+    const actionUser = users.find((item) => item.id === actionUserId);
 
     return (
       <div>
-        <PageHeader text={PAGE_TITLES.members} />
+        <PageHeader text={PAGE_TITLES.members} onClick={() => this.setModalMode(MODAL_MODES.create)} />
```
Comment (maximsan) [inline src/pages/members/Members.jsx:105]: create handler

File: src/pages/members/Members.jsx [hunk 11]
```diff
@@ -79,23 +123,35 @@ export class Members extends PureComponent {
...
             })}
           </tbody>
         </table>
-        {deleteMode && (
+        {modalMode === MODAL_MODES.delete && (
           <DeleteModal
             target={DELETE_VALUES.member}
             removeHandler={this.removeUser}
-            disableModalMode={this.disableDeleteMode}
+            disableModalMode={this.disableModalMode}
```
Comment (maximsan) [inline src/pages/members/Members.jsx:143]: `close` or `onClose` if it closes modal

File: src/pages/members/Members.jsx [hunk 12]
```diff
@@ -79,23 +123,35 @@ export class Members extends PureComponent {
...
                 />
               );
             })}
           </tbody>
         </table>
-        {deleteMode && (
+        {modalMode === MODAL_MODES.delete && (
           <DeleteModal
             target={DELETE_VALUES.member}
             removeHandler={this.removeUser}
```
Comment (maximsan) [inline src/pages/members/Members.jsx:142]: `onRemove` or `onDelete`

**Review by maximsan — CHANGES_REQUESTED** [review]
Please remove package-lock.json file, reinstall all packages, fix notes. Thanks!

File: src/pages/modals/form/formField/FormField.jsx [hunk 13]
```diff
@@ -1,25 +1,45 @@
...
+              </option>
+            ))}
+          </select>
+        ) : (
+          <input
+            type={stylingType}
+            disabled={readOnly}
+            name={inputName}
+            className={styles.input}
+            placeholder={fieldName}
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/form/formField/FormField.jsx:38]: Fixed

File: src/pages/modals/form/formField/FormField.jsx [hunk 14]
```diff
@@ -1,25 +1,45 @@
...
+        {selectValues ? (
+          <select
+            value={inputValue}
+            className={inputValue ? styles.input : `${styles.input} ${styles.placeholder}`}
+            disabled={readOnly}
+            name={inputName}
+            onChange={changeInputHandler}
+          >
+            <option value='' disabled hidden>
+              {fieldName}
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/form/formField/FormField.jsx:24]: Fixed

File: src/pages/modals/form/formField/FormField.jsx [hunk 15]
```diff
@@ -1,25 +1,45 @@
 import PropTypes from 'prop-types';
 import styles from './FormField.module.css';
 import noop from '../../../../shared/noop';
+import { INPUT_TYPES } from '../../../../constants/libraries';
 
-export function FormField({ onChange, placeholder, inputValue, fieldName, readOnly, stylingType, inputName, error }) {
-  const onChangeHandler = ({ target: { name, value } }) => {
+export function FormField({ onChange, inputValue, fieldName, readOnly, stylingType, inputName, error, selectValues }) {
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/form/formField/FormField.jsx:6]: Created 2 components, `<Input/>` and `<Select/>`

File: src/pages/modals/taskModals/taskForm/TaskModalUsersList.jsx [hunk 16]
```diff
@@ -9,14 +9,24 @@ export function TaskModalUsersList({ usersTask, changeUserValue, error, readOnly
...
                   changeUserValue(user.id, event.currentTarget.checked);
                 };
 
                 return (
-                  <Checkbox key={user.id} value={user.value} onChange={onChangeHandler} text={user.name} id={user.id} />
+                  <Checkbox
+                    key={user.id}
+                    value={user.value}
+                    onChange={onChangeHandler}
+                    text={`${user.name} ${user.surname}`}
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/taskModals/taskForm/TaskModalUsersList.jsx:27]:
```suggestion
export const getFullName = (name, surname) => {
  return `${name} ${surname}`;
};
```
Fixed

File: src/pages/modals/userModal/helperTests/getReadUserModalState.test.js (hunk 1 above)
Comment (oleg3555) [reply src/pages/modals/userModal/helperTests/getReadUserModalState.test.js:71]: Fixed

File: src/pages/modals/userModal/userModalHelpers.js (hunk 2 above)
Comment (oleg3555) [reply src/pages/modals/userModal/userModalHelpers.js:24]: Fixed

File: src/pages/modals/userModal/userModalHelpers.js [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+  skype: '',
+  startDate: '',
+  education: '',
+  averageScore: '',
+  mathScore: '',
+};
+
+export const initStartUserModalState = {
+  ...formFields,
+  modalTitle: '',
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/userModal/userModalHelpers.js:26]: Fixed

File: src/pages/modals/userModal/userModalHelpers.js [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+  readOnly: false,
+  formErrors: {
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/userModal/userModalHelpers.js:28]: Fixed

File: src/pages/modals/userModal/userModalHelpers.js [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+    ...formFields,
+  },
+};
+
+export const getUserModalErrors = (state) => {
+  const { email, password, confirmPassword, birthDate, phone } = state;
+  const formErrors = { ...formFields };
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/userModal/userModalHelpers.js:35]: Fixed

File: src/pages/modals/userModal/userModalHelpers.js [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+  if (!isObjectFieldsEmpty(formErrors)) {
+    return formErrors;
+  }
+
+  return null;
+};
+
+export const getReadUserModalState = (user) => {
+  return { ...initStartUserModalState, ...user, modalTitle: USER_MODAL_TITLES.read, readOnly: true };
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/userModal/userModalHelpers.js:69]: Fixed

File: src/pages/modals/userModal/userModalHelpers.js (hunk 3 above)
Comment (oleg3555) [reply src/pages/modals/userModal/userModalHelpers.js:78]:
```suggestion
const getModalTitle = (user, readOnly) => {
  if (readOnly && user) {
    return USER_TITLES.read;
  }
  if (user) {
    return USER_TITLES.edit;
  }

  return USER_TITLES.read;
};

export const getUserModalState = (user = null, readOnly = false) => {
  const title = getModalTitle(user, readOnly);

  return { ...userModalState, ...user, title, readOnly };
};
```
And remove 3 mrthods from UserModal.jsx and looks better, thanks
```suggestion
componentDidMount() {
    const { user, readOnly } = this.props;
    const data = getUserModalState(user, readOnly);
    this.setState(data);
  }
```

File: src/pages/modals/userModal/userModalHelpers.js [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+
+export const getModalUserData = (state) => {
+  const keys = Object.keys(formFields);
+  const data = {};
+  keys.forEach((key) => {
+    data[key] = state[key].trim();
+  });
+
+  return data;
+};
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/modals/userModal/userModalHelpers.js:88]: Rewatched my other modal helpers, and I'm not reusing it

File: src/pages/setPassword/SetPassword.jsx (hunk 6 above)
Comment (oleg3555) [reply src/pages/setPassword/SetPassword.jsx:74]: Done)

File: src/tests/getAge.test.js (hunk 8 above)
Comment (oleg3555) [reply src/tests/getAge.test.js:3]: Fixed

File: src/tests/getUserName.test.js (hunk 9 above)
Comment (oleg3555) [reply src/tests/getUserName.test.js:3]: Fixed

File: src/pages/members/Members.jsx (hunk 10 above)
Comment (oleg3555) [reply src/pages/members/Members.jsx:105]:
```suggestion
<PageHeader text={PAGE_TITLES.members} onClick={this.setModalMode} />
```
And in setModalMode I set it as default parametr
```suggestion
setModalMode = (modalMode = MODAL_MODES.create, actionUserId = null) => {
    this.setState({ modalMode, actionUserId });
  };
```
Is it ok?

File: src/pages/tasks/Tasks.jsx (hunk 7 above)
Comment (oleg3555) [reply src/pages/tasks/Tasks.jsx:58]: Fixed

File: src/pages/members/Members.jsx (hunk 12 above)
Comment (oleg3555) [reply src/pages/members/Members.jsx:142]: Fixed

File: src/pages/setPassword/SetPassword.jsx [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+    this.state = {
+      user: null,
+      password: '',
+      error: false,
+    };
+  }
+
+  async componentDidMount() {
+    const cryptedId = getUid(document.location.search);
+    const id = encryptId(cryptedId);
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/setPassword/SetPassword.jsx:24]: const id = getUid(document.location.search);
And in helper
```suggestion
export function getUid(url) {
  const idParameter = 'uid%3D';
  const cryptedId = url.substring(url.indexOf(idParameter) + idParameter.length, url.indexOf('&lang'));

  return encryptId(cryptedId);
}
```
Is  it ok?

File: src/pages/setPassword/SetPassword.jsx (hunk 4 above)
Comment (oleg3555) [reply src/pages/setPassword/SetPassword.jsx:43]: Created method
```suggestion
export const isPasswordValid = (password) => {
  return password.length >= 8 && password.length <= 24;
};
```

File: src/pages/setPassword/SetPassword.jsx (hunk 5 above)
Comment (oleg3555) [reply src/pages/setPassword/SetPassword.jsx:59]: Fixed

File: src/pages/members/Members.jsx (hunk 12 above)
Comment (oleg3555) [reply src/pages/members/Members.jsx:142]: Fixed

File: src/pages/members/Members.jsx (hunk 11 above)
Comment (oleg3555) [reply src/pages/members/Members.jsx:143]: Fixed

File: src/pages/setPassword/SetPassword.jsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/pages/setPassword/SetPassword.jsx]: `export function getUid(url) {
  const cryptedId = url.match(cryptedIdRegular).pop();

  return encryptId(cryptedId);
}`

`cryptedIdRegular = /[0-9]+(a[0-9]+)+/g;`

Fixed

**Review by maximsan — APPROVED** [review]
Looks fine. I left a comment. Don't forget to add tests for methods you forgot to cover. It's approved 🚀

**Review by maximsan — CHANGES_REQUESTED** [review]
Pardon, but where is theme switcher ? Please add a user into pr description I can use to login

**Review by maximsan — APPROVED** [review]

**Pull request merged**